text/subsingleton-elimination.md

```diff
@@ -0,0 +1,176 @@
+
+There are two solutions:
+- either we retarget `Prop` to include `HProp` (simpler)
+- or we keep `Prop` as it is and add a new subuniverse `HProp` (less change of habits)
+
+Both solutions follow the same way. Below, we assume the latter. To get the former, just use `Prop` instead of `HProp`.
+
+We shall avoid relying on template polymorphism. However, we may consider a `cast` of the form:
+```
+(t : HProp by p)
+```
+which forces the type-checker to recognize a type as being in a smaller universe than decidably detectable.
+
+The status of casts in the implementation is however fragile, so this may require some clarification of casts first.
+
+## New command to support arbitrary `HProp`
+
+We propose a new command:
+```
+Subsingleton id context : ishProp A := proof.
+```
+which recast the type of `A` from `context -> Type` to `context -> HProp` so that it is latter considered impredicative.
+
+An alternative is to have a modifier to definitions in the form:
+```
+Subsingleton id context := def : HProp by proof.
+```
+But then, we probably want also:
+```
+Subsingleton id context := def : Hprop.
+Proof.
+the proof of being hProp
+Qed.
+```
+which may be complicated.
+
+If `HProp` is made distinct from `Prop`, new declarations `Inductive ... : ... -> HProp` would be accepted too. When to tell that the type has subsingleton elimination? There is a proposal in the next section.
+
+## Dynamic detection of subsingleton elimination
+
+We propose an extension of `match` of the following form:
+```
+match t as id in I params return typ by u with
+| C1 vars => body
+...
+| Cn vars => body
+end
+```
+where, when `I` is `HProp`-truncated but the instance `I args` is an `HProp` proved by `u : forall x y:I args, x = y`, the elimination to `Type` is allowed.
+
+# Side remark on template polymorphism for `Prop`
+
+Instances of parametric types which are not in `Prop` in the general case are automatically recognized as belonging to the `Prop` class (and treated as subsingleton for `match` and as impredicative in function types) when the parameters are themselves in `Prop`. This is typically the case of `prod A B` which behaves as `and A B` when `A` and `B` are in `Prop`.
+
+Template polymorphism induces a view where an inductive type in `Prop` does not necessarily mean explicitly `Prop`-truncated. It is so only when the original type is not a syntactic `Prop`-subsingleton.
+
+Without template polymorphism, explicit coercions from `prod A B` to `and A B` should be inserted (when `A` and `B` are `Prop`), and explicit Prop-boxing of `prod A B` should be done to make `prod A B` behave impredicatively.
+
+# Issues with extraction of `HProp`
+
+Types in `HProp` which are not in the current syntactic class `Prop` could simply be extracted as if they were types.
+
+# Conclusion
+
+A few extensions could be done pretty easily:
+1. accept singleton types with all arguments in `SProp` to be in `SProp` (`True`, `and`, ...)
```

Comment:
This is not especially useful as we already have records, and is not trivial to implement as the reduction rule for the match is something like
~~~
match p with conj x y => e end 
====>
e[x := match p with conj x _ => x end; y := match p with conj _ y => y end]
~~~
Maybe I'm overestimating the difficulty but I don't think I'm underestimating the usefulness.

Reply:
>  I don't think I'm underestimating the usefulness.

Indeed, I forgot eta for non empty primitive records.

(Then, aren't we back to the practical question of letting existing libraries take benefit of primitive records?)

> Maybe I'm overestimating the difficulty

Isn't there two levels?

Couldn't `and` be in `SProp` even with the usual (non proof-irrelevant) `match` reduction?

But more fundamentally, you're right that the good answer is probably more about stopping to distinguish between `match`-based `and` and projection-based `and` (or, if we don't unify them, to at least get rid of `match`-based `and).

In particular, it is a bit misleading that `and` is accepted in `SProp` without being recognized as subsingleton. It should at least be discouraged.

Reply:
>Couldn't and be in SProp even with the usual (non proof-irrelevant) match reduction?

That would break decidability of typechecking (same as equality without reducing on reflexive proofs)

Reply:
> > Couldn't and be in SProp even with the usual (non proof-irrelevant) match reduction?
> 
> That would break decidability of typechecking (same as equality without reducing on reflexive proofs)

I'm sorry, I fail to see the example. (By the way, do you have a copy of your PhD somewhere?)

Also, I added to the CEP a section on `SProp` subtype of `Prop`. Did I miss an issue with decidability?

Reply:
How do you check
~~~coq 
fun (x : and True True) (f : match x with conj _ _ => nat -> nat end) => f 0
~~~
?

(specifically it's not that it's undecidable, it's that I don't know how to decide it without the reduction)

Reply:
Thanks. So what is the consensus about this situation? That we should reduce `match` of positive records as you gave above? Or that we should get rid of positive records at all?